File: src/main/java/com/movieapp/movieapp/service/MovieService.java
package com.movieapp.movieapp.service;

import com.movieapp.movieapp.entity.Movie;
import com.movieapp.movieapp.repository.MovieRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class MovieService {

    private final MovieRepository movieRepository;
    private final RestTemplate restTemplate;

    @Value("${omdb.api.url}")
    private String omdbUrl;

    @Value("${omdb.api.key}")
    private String omdbKey;

    public MovieService(MovieRepository movieRepository, RestTemplate restTemplate) {
        this.movieRepository = movieRepository;
        this.restTemplate = restTemplate;
    }

    public Page<Movie> getAllMovies(Pageable pageable) {
        return movieRepository.findAll(pageable);
    }

    public Optional<Movie> getMovieById(Long id) {
        return movieRepository.findById(id);
    }

    public Movie addMovie(Movie movie) {
        if (movieRepository.existsByImdbId(movie.getImdbId())) {
            throw new RuntimeException("Movie with IMDb ID " + movie.getImdbId() + " already exists");
        }
        return movieRepository.save(movie);
    }

    public List<Movie> addBatch(List<Movie> movies) {
        List<Movie> added = new ArrayList<>();
        for (Movie movie : movies) {
            if (!movieRepository.existsByImdbId(movie.getImdbId())) {
                added.add(movieRepository.save(movie));
            }
        }
        return added;
    }

    public Optional<Movie> updateMovie(Long id, Movie updatedMovie) {
        return movieRepository.findById(id).map(existing -> {
            existing.setTitle(updatedMovie.getTitle());
            existing.setYear(updatedMovie.getYear());
            existing.setImdbId(updatedMovie.getImdbId());
            existing.setType(updatedMovie.getType());
            existing.setPosterUrl(updatedMovie.getPosterUrl());
            existing.setGenre(updatedMovie.getGenre());
            existing.setDirector(updatedMovie.getDirector());
            existing.setActors(updatedMovie.getActors());
            existing.setPlot(updatedMovie.getPlot());
            return movieRepository.save(existing);
        });
    }

    public boolean deleteMovie(Long id) {
        if (movieRepository.existsById(id)) {
            movieRepository.deleteById(id);
            return true;
        }
        return false;
    }

    public void deleteBatch(List<Long> ids) {
        movieRepository.deleteAllById(ids);
    }

    public List<Movie> searchDb(String query) {
        return movieRepository.findByTitleContainingIgnoreCase(query);
    }

    public List<Movie> searchOmdb(String query) {
        try {
            String apiUrl = omdbUrl + "?s=" + URLEncoder.encode(query, StandardCharsets.UTF_8) + "&apikey=" + omdbKey;
            @SuppressWarnings("unchecked")
            Map<String, Object> response = restTemplate.getForObject(apiUrl, Map.class);
            if (response == null || "False".equals(response.get("Response"))) {
                throw new RuntimeException((String) response.getOrDefault("Error", "OMDB search failed"));
            }
            @SuppressWarnings("unchecked")
            List<Map<String, String>> searchResults = (List<Map<String, String>>) response.get("Search");
            return searchResults.stream()
                    .map(result -> Movie.builder()
                            .title(result.get("Title"))
                            .year(result.get("Year"))
                            .imdbId(result.get("imdbID"))
                            .type(result.get("Type"))
                            .posterUrl(result.get("Poster"))
                            .build())
                    .collect(Collectors.toList());
        } catch (Exception e) {
            throw new RuntimeException("OMDB API error: " + e.getMessage());
        }
    }

    // ✅ New Method: fetch detailed movie info from OMDB
    public Movie fetchMovieDetails(String imdbId) {
        try {
            String apiUrl = omdbUrl + "?i=" + imdbId + "&apikey=" + omdbKey;
            @SuppressWarnings("unchecked")
            Map<String, String> response = restTemplate.getForObject(apiUrl, Map.class);
            if (response == null || "False".equals(response.get("Response"))) {
                throw new RuntimeException("Movie not found");
            }
            return Movie.builder()
                    .title(response.get("Title"))
                    .year(response.get("Year"))
                    .imdbId(response.get("imdbID"))
                    .type(response.get("Type"))
                    .posterUrl(response.get("Poster"))
                    .genre(response.get("Genre"))
                    .director(response.get("Director"))
                    .actors(response.get("Actors"))
                    .plot(response.get("Plot"))
                    .build();
        } catch (Exception e) {
            throw new RuntimeException("OMDB API error: " + e.getMessage());
        }
    }

    // ✅ New Method: fetch and save movies from OMDB
    public List<Movie> fetchAndSaveFromOmdb(String query) {
        List<Movie> fetchedMovies = searchOmdb(query); // use existing OMDB search
        List<Movie> savedMovies = new ArrayList<>();

        for (Movie movie : fetchedMovies) {
            if (!movieRepository.existsByImdbId(movie.getImdbId())) {
                // Fetch detailed info before saving
                Movie detailedMovie = fetchMovieDetails(movie.getImdbId());
                savedMovies.add(movieRepository.save(detailedMovie));
            }
        }

        return savedMovies;
    }

    public Map<String, Long> getStats() {
        Map<String, Long> stats = new HashMap<>();
        stats.put("totalMovies", movieRepository.count());
        return stats;
    }
}

File: movie-frontend/src/app/core/movie.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../environments/environment';
import { Observable, throwError } from 'rxjs';
import { AuthService } from './auth.service';

export interface Movie {
  id?: number;
  title: string;
  year?: string;
  imdbId?: string;
  posterUrl?: string;
  plot?: string;
  type?: string;
  genre?: string;
  director?: string;
  actors?: string;
}

export interface Page<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

@Injectable({ providedIn: 'root' })
export class MovieService {
  constructor(private http: HttpClient, private authService: AuthService) {}

  getAll(page = 0, size = 10): Observable<Page<Movie>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<Page<Movie>>(`${environment.apiBaseUrl}/api/movies`, { params });
  }

  getById(id: number): Observable<Movie> {
    return this.http.get<Movie>(`${environment.apiBaseUrl}/api/movies/${id}`);
  }

  searchDb(query: string): Observable<Movie[]> {
    return this.http.get<Movie[]>(`${environment.apiBaseUrl}/api/movies/search`, {
      params: { query }
    });
  }

  searchOmdb(query: string): Observable<Movie[]> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can search OMDB'));
    }
    return this.http.get<Movie[]>(`${environment.apiBaseUrl}/api/movies/search-omdb`, {
      params: { query }
    });
  }

  add(movie: Movie): Observable<Movie> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can add movies'));
    }
    return this.http.post<Movie>(`${environment.apiBaseUrl}/api/movies`, movie);
  }

  addBatch(movies: Movie[]): Observable<Movie[]> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can add movies'));
    }
    return this.http.post<Movie[]>(`${environment.apiBaseUrl}/api/movies/batch`, movies);
  }

  fetchAndSave(query: string): Observable<Movie[]> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can fetch and save movies'));
    }
    return this.http.post<Movie[]>(`${environment.apiBaseUrl}/api/movies/fetch`, null, {
      params: { query }
    });
  }

  update(id: number, movie: Movie): Observable<Movie> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can update movies'));
    }
    return this.http.put<Movie>(`${environment.apiBaseUrl}/api/movies/${id}`, movie);
  }

  delete(id: number): Observable<void> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can delete movies'));
    }
    return this.http.delete<void>(`${environment.apiBaseUrl}/api/movies/${id}`);
  }

  deleteBatch(ids: number[]): Observable<void> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can delete movies'));
    }
    return this.http.post<void>(`${environment.apiBaseUrl}/api/movies/delete-batch`, ids);
  }

  getStats(): Observable<{totalMovies: number}> {
    if (this.authService.getUserRole() !== 'ADMIN') {
      return throwError(() => new Error('Unauthorized: Only admins can view movie stats'));
    }
    return this.http.get<{totalMovies: number}>(`${environment.apiBaseUrl}/api/movies/stats`);
  }
}



